src/citrine/informatics/design_spaces/hierarchical_design_space.py

```diff
@@ -0,0 +1,215 @@
+        self.template_link: Optional[TemplateLink] = template_link
+        self.display_name: Optional[str] = display_name
+
+    def __repr__(self):
+        display_name = self.display_name or self.name
+        return "<MaterialNodeDefinition {!r}>".format(display_name)
+
+
+class HierarchicalDesignSpace(ModuleEngineResource["HierarchicalDesignSpace"], DesignSpace):
+    """A design space that produces hierarchical candidates representing a material history.
+
+    A hierarchical design space always contains a root node that defines the
+    attributes and formulation contents included on terminal materials of the candidates.
+    It also includes a set of sub-nodes that can be used to define the any new
+    materials that appear in the history of the terminal material.
+
+    Material histories produced by this design space are connected based on the
+    name identifiers and formulation contents of each node.
+    For example, the root node may contain a formulation subspace with an ingredient named
+    'New Mixture-001'. If this ingredient matches the `name` field on one of the sub-nodes
+    in the design space, the resulting candidates will contain a terminal material with this
+    new mixture as one of its ingredients. This procedure can be extended to sub-nodes
+    referencing other sub-nodes, allowing for the linkage of complex material history shapes
+    in the resulting candidates.
+
+    Every node also contains a set of :class:`~citrine.informatics.dimensions.Dimension`s
+    used to define any attributes (i.e., properties, processing parameters)
+    that will appear on the materials produced by that node.
+
+    :class:`~citrine.informatics.data_sources.DataSource`s can be included on the configuration
+    to allow for design over "known" materials. The Citrine Platform will look up
+    the any ingredient names from formulation subspaces on the design space nodes
+    in order to inject their composition/properties into the material history of the candidates.
+    When constructing a default hierarchical design space,
+    the Citrine Platform includes any data sources found on the provided predictor configuration.
+
+    Parameters
+    ----------
+    name: str
+        Name of the design space
+    description: str
+        Description of the design space
+    root: MaterialNodeDefinition
+        Terminal material node produced by the design space
+    subspaces: List[MaterialNodeDefinition]
+        Sub material nodes produced by the design space
+    data_sources: List[DataSource]
+        Data sources used to inject known materials into the hierarchical candidates
+
+    """
+
+    root = properties.Object(MaterialNodeDefinition, "data.instance.root")
+    subspaces = properties.List(
+        properties.Object(MaterialNodeDefinition), "data.instance.subspaces", default=list()
+    )
+    data_sources = properties.List(
+        properties.Object(DataSource), "data.instance.data_sources", default=list()
+    )
```

Comment:
As an FYI, by setting the default value to a particular list, all class instances that use that default value.  You actually share a single mutable object between multiple objects, and edits to one containing object will happen to all.  This is probably best read as a bug in our serialization stack.  

Just so you know.

Reply:
I think this is just a bug in our serialization library? We do it all over the place in here.

Reply:
We don't even need a default. The backend has these fields as non optional but with empty list defaults, so the field will always be included